File: aaaTercero/PREPFINAL/Febrero 2019/feb2019haskell/Kruskal.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME:
-- GRADO/STUDIES:
-- NÚM. MÁQUINA/MACHINE NUMBER:
--
----------------------------------------------

module Kruskal(kruskal, kruskals) where

import qualified DataStructures.Dictionary.AVLDictionary as D
import qualified DataStructures.PriorityQueue.LinearPriorityQueue as PQ
import DataStructures.Graph.DictionaryWeightedGraph

kruskal :: (Ord a, Ord w) => WeightedGraph a w -> [WeightedEdge a w]
kruskal g = aux (diccionrep (vertices g)) (crearcola (edges g)) []
    where
        aux dict p l 
            |PQ.isEmpty p = l
            |(representante vs dict) /= (representante vd dict)= aux (D.insert (representante vd dict) vs dict) (PQ.dequeue p) (((PQ.first p)):l)
            |otherwise = aux dict (PQ.dequeue p) l
                where
                    (WE vs w vd) = PQ.first p

crearcola::(Ord a, Ord w)=>[WeightedEdge a w]->PQ.PQueue (WeightedEdge a w)
crearcola l = foldr (PQ.enqueue) PQ.empty l

diccionrep::(Ord a)=>[a] -> D.Dictionary a a
diccionrep l= auxdl l D.empty
    where
        auxdl [] diccrep = diccrep
        auxdl (y:ys) diccrep = auxdl ys (D.insert y y diccrep)

representante::(Ord a)=>a -> D.Dictionary a a ->a
representante v dict 
    |v == valor = v
    |otherwise = representante valor dict
        where 
            Just valor = D.valueOf v dict 

-- Solo para evaluación continua / only for part time students
kruskals :: (Ord a, Ord w) => WeightedGraph a w -> [[WeightedEdge a w]]
kruskals = undefined

File: TERCERO/PREPARACION PARA EL FINAL/Febrero 2019/java/src/Kruskal.java
/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME:
 * -- GRADO/STUDIES:
 * -- NÚM. MÁQUINA/MACHINE NUMBER:
 * --
 * ----------------------------------------------
 */

import dataStructures.graph.DictionaryWeightedGraph;
import dataStructures.graph.WeightedGraph;
import dataStructures.graph.WeightedGraph.WeightedEdge;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.priorityQueue.PriorityQueue;
import dataStructures.priorityQueue.LinkedPriorityQueue;
import dataStructures.set.Set;
import dataStructures.set.HashSet;

public class Kruskal {
	public static <V,W> Set<WeightedEdge<V,W>> kruskal(WeightedGraph<V,W> g) {
		// COMPLETAR
		Set<WeightedEdge<V,W>> s= new HashSet<>();
		PriorityQueue<WeightedEdge<V,W>> PQ = new LinkedPriorityQueue<>();
		Dictionary<V,V> dicrep = new HashDictionary<>();
		for(V v : g.vertices()) {
			dicrep.insert(v,v);
		}
		for(WeightedEdge<V,W> w : g.edges()) {
			PQ.enqueue(w);
		}
		WeightedEdge<V,W> w;
		while(!PQ.isEmpty()) {
			w = PQ.first();
			PQ.dequeue();
			if(comprobarrep(dicrep,w.source()) != comprobarrep(dicrep, w.destination())){
				dicrep.insert(comprobarrep(dicrep, w.destination()), w.source());
				s.insert(w);
			}
		}
		return s;
	}

	private static <V>V comprobarrep(Dictionary<V,V> dict, V ver) {
		while(dict.valueOf(ver) != ver) {
			ver = dict.valueOf(ver);
		}
		return  ver;
	}

	// Sólo para evaluación continua / only for part time students
	public static <V,W> Set<Set<WeightedEdge<V,W>>> kruskals(WeightedGraph<V,W> g) {

		// COMPLETAR
		
		return null;
	}
}
